db/dcrsqlite/sqlite.go

```diff
@@ -929,6 +1074,120 @@ func (db *DB) RetrieveStakeInfoExtended(ind int64) (*apitypes.StakeInfoExtended,
+func (db *DB) JustifyTableStructures(dbInfo *DBInfo) error {
+
+	// Grab the column info. Right now, just counting them, but could be looked at more closely
+	// Each tuple is a row with columns [cid, name, type, notnull, dflt_value, pk]
+	rows, err := db.Query(db.getBlockSummaryTableInfo)
+	if err != nil {
+		log.Errorf("Query failed: %v", err)
+		return err
+	}
+	defer rows.Close()
+	rowCounter := 0
+	for rows.Next() {
+		rowCounter += 1
+	}
+
+	// It simply checks whether there are enough columns for the current structure
+	if rowCounter >= 12 {
+		return nil
+	}
+
+	log.Info("Detected old SQLite table structure. Updating.")
+
+	// Create a backup file, if one hasn't already been created.
+	directory := filepath.Dir(dbInfo.FileName)
+	bkpPath := filepath.Join(directory, "dcrdata.nosidechains-bkp.db")
+	_, err = os.Stat(bkpPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			err = copyFile(dbInfo.FileName, bkpPath)
+			if err != nil {
+				log.Errorf("Failed to backup %s: %v", dbInfo.FileName, err)
+				return err
+			}
+		} else {
+			log.Errorf("Error retrieving FileInfo for %s: %v", dbInfo.FileName, err)
+			return err
+		}
+	}
+
+	tmpSummaryTableName := TableNameSummaries + "_temp"
+	tmpStakeTableName := TableNameStakeInfo + "_temp"
+
+	queries := make([]string, 0, 8)
+	queries = append(queries, fmt.Sprintf(db.rawCreateBlockSummaryStmt, tmpSummaryTableName))
+	queries = append(queries, fmt.Sprintf(`INSERT INTO %s SELECT hash, height, size, diff,
+		sdiff, time, poolsize, poolval, poolavg, winners, 1, 1 FROM %s;`,
+		tmpSummaryTableName, TableNameSummaries))
+	queries = append(queries, fmt.Sprintf("DROP TABLE %s;", TableNameSummaries))
+	queries = append(queries, fmt.Sprintf("ALTER TABLE %s RENAME TO %s;",
+		tmpSummaryTableName, TableNameSummaries))
+	queries = append(queries, fmt.Sprintf(db.rawCreateStakeInfoExtendedStmt, tmpStakeTableName))
+	queries = append(queries, fmt.Sprintf(`INSERT INTO %[1]s SELECT %[2]s.hash, %[3]s.height, num_tickets,
+		fee_min, fee_max, fee_mean, fee_med, fee_std, %[3]s.sdiff, window_num, window_ind, pool_size,
+		pool_val,pool_valavg, %[3]s.winners FROM %[3]s JOIN %[2]s ON
+		%[3]s.height = %[2]s.height`,
+		tmpStakeTableName, TableNameSummaries, TableNameStakeInfo))
+	queries = append(queries, fmt.Sprintf("DROP TABLE %s;", TableNameStakeInfo))
+	queries = append(queries, fmt.Sprintf("ALTER TABLE %s RENAME TO %s;", tmpStakeTableName, TableNameStakeInfo))
+
+	transaction, err := db.Begin()
+	if err != nil {
+		log.Errorf("Failed to start a transaction: \n", err)
+		return err
+	}
+	for _, query := range queries {
+		_, err = transaction.Exec(query)
+		if err != nil {
+			log.Errorf("Failed updating SQLite table structure: \n", err)
+			transaction.Rollback()
+			return err
+		}
+	}
+
+	err = transaction.Commit()
+	if err != nil {
+		log.Error("Failed to commit SQL transaction after table reorg.")
+		return err
+	}
+
+	// Clean up the file a little bit
+	_, err = db.Query("VACUUM;")
```

Comment:
This needs to be `_, err = db.Exec("VACUUM;")`

Reply:
After the table upgrade with no VACUUM, the files size was 451538944 B.
With VACUUM, the file size is reduced to 323424256 B.